neqo-transport/src/server.rs

```diff
@@ -317,20 +301,22 @@ impl Server {
             remote_address: dgram.source(),
             odcid: orig_dcid.as_ref().unwrap_or(&initial.dst_cid).clone(),
         };
-        let connection = self.connections.borrow().values().find_map(|c| {
-            (c.borrow().active_attempt.as_ref() == Some(&attempt_key)).then(|| Rc::clone(c))
-        });
-        connection.map_or_else(
-            || self.accept_connection(&attempt_key, initial, dgram, orig_dcid, now),
-            |c| {
-                qdebug!(
-                    [self],
-                    "Handle Initial for existing connection attempt {:?}",
-                    attempt_key
-                );
-                c.borrow_mut().process(Some(dgram), now)
-            },
-        )
+
+        let Some(connection) = self
+            .connections
+            .iter_mut()
+            .find(|c| c.active_attempt.as_ref() == Some(&attempt_key))
+        else {
+            return self.accept_connection(&attempt_key, initial, dgram, orig_dcid, now);
+        };
+
+        let out = connection.process(Some(dgram), now);
+        qdebug!(
+            [self],
+            "Handled Initial for existing connection attempt {:?}",
+            attempt_key
+        );
+        out
```

Comment:
We might want to add a test for this error case.

Reply:
I looked into this. With the above change, [namely that two connection attempts with the same original destination connection id but different paths, result in a single connection](https://github.com/mozilla/neqo/pull/1981#discussion_r1676370725), I believe this code path is dead.

A second initial (i.e. an initial for an existing connection attempt) will be passed to the existing connection in `Server::process_input` right away:

https://github.com/mozilla/neqo/blob/8b81a090384076333f4c5dd54b2983d5da292daa/neqo-transport/src/server.rs#L502-L505

I believe this makes the entire concept of `AttemptKey` redundant. Am I missing something?

https://github.com/mozilla/neqo/blob/8b81a090384076333f4c5dd54b2983d5da292daa/neqo-transport/src/server.rs#L82-L90

Thanks for raising this @larseggert!

Reply:
If it's dead, remove and maybe add a `debug_assert`?

Reply:
:+1: done in https://github.com/mozilla/neqo/pull/1981/commits/05442f58e046edffaa8f5f8a999570cf969ca0af.